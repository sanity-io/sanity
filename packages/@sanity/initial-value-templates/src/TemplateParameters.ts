export type TemplateParameter = FieldDefinition | ArrayFieldDefinition

export interface TypeTarget {
  type: string
}

export interface ReferenceTarget {
  type: 'reference'
  to: TypeTarget | TypeTarget[]
}

export interface FieldDefinition {
  name: string
  type: string
  title?: string
  description?: string
  options?: {[key: string]: any}
}

export type ReferenceFieldDefinition = FieldDefinition & ReferenceTarget

export type ArrayFieldDefinition = FieldDefinition & {
  type: 'array'
  of: (ReferenceTarget | TypeTarget)[]
}
