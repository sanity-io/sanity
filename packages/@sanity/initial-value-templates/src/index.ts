export {default as TemplateBuilder} from './builder'
export {resolveInitialValue, isBuilder} from './resolve'
export {validateTemplates} from './validate'
export {
  templateExists,
  getTemplates,
  getTemplateById,
  getTemplatesBySchemaType,
  getParameterlessTemplatesBySchemaType,
  getTemplateErrors
} from './templates'
