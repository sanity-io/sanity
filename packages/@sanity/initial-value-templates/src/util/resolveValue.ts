import {InitialValueParams, InitialValueProperty, InitialValueResolver} from '@sanity/types'

// returns the "resolved" value from an initial value property (e.g. type.initialValue)
export async function resolveValue<InitialValue>(
  initialValueOpt: InitialValueProperty<InitialValue>,
  params?: InitialValueParams
): Promise<InitialValue | undefined> {
  return typeof initialValueOpt === 'function'
    ? (initialValueOpt as InitialValueResolver<InitialValue>)(params)
    : initialValueOpt
}
