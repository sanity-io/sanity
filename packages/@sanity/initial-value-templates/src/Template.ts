import {TemplateParameter} from './TemplateParameters'

type ValueResolver = (parameters: {[key: string]: any}) => {[key: string]: any}

export interface Template {
  id: string
  title: string
  description?: string
  schemaType: string
  icon?: Function
  value: ValueResolver | {[key: string]: any}
  parameters?: TemplateParameter[]
}

export class TemplateBuilder {
  spec: Partial<Template>

  constructor(spec?: Template) {
    this.spec = spec || {}
  }

  id(id: string) {
    return this.clone({id})
  }

  getId() {
    return this.spec.id
  }

  title(title: string) {
    return this.clone({title})
  }

  description(description: string) {
    return this.clone({description})
  }

  getDescription() {
    return this.spec.description
  }

  getTitle() {
    return this.spec.title
  }

  schemaType(typeName: string) {
    return this.clone({schemaType: typeName})
  }

  getSchemaType() {
    return this.spec.schemaType
  }

  icon(icon: Function) {
    return this.clone({icon})
  }

  getIcon() {
    return this.spec.icon
  }

  value(value: ValueResolver | {[key: string]: any}) {
    return this.clone({value})
  }

  getValue() {
    return this.spec.value
  }

  parameters(parameters: TemplateParameter[]) {
    return this.clone({parameters})
  }

  getParameters() {
    return this.spec.parameters
  }

  serialize(): Template {
    const {id, title, description, schemaType, value, icon, parameters} = this.spec
    if (!id) {
      throw new Error('Template is missing required "id"')
    }

    if (!title) {
      throw new Error(`Template with ID "${id}" is missing required "title"`)
    }

    if (!schemaType) {
      throw new Error(`Template with ID "${id}" is missing required "schemaType"`)
    }

    if (!value) {
      throw new Error(`Template with ID "${id}" is missing required "value"`)
    }

    return {id, title, description, schemaType, value, icon, parameters}
  }

  clone(withSpec?: Partial<Template>) {
    const builder = new TemplateBuilder()
    builder.spec = {...this.spec, ...(withSpec || {})}
    return builder
  }
}
