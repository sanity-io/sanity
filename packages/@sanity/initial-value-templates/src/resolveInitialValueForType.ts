import {isEmpty, resolveTypeName} from '@sanity/util/content'

import {
  ArraySchemaType,
  InitialValueParams,
  isArraySchemaType,
  isObjectSchemaType,
  ObjectSchemaType,
  SchemaType,
} from '@sanity/types'

import {randomKey} from '@sanity/util/paths'
import deepAssign from './util/deepAssign'
import {resolveValue} from './util/resolveValue'

export function getItemType(arrayType: ArraySchemaType, item: unknown): SchemaType | undefined {
  const itemTypeName = resolveTypeName(item)

  return itemTypeName === 'object' && arrayType.of.length === 1
    ? arrayType.of[0]
    : arrayType.of.find((memberType) => memberType.name === itemTypeName)
}

const MAX_RECURSION_DEPTH = 10

/**
 * Resolve initial value for the given schema type (recursively)
 *
 * @param type {SchemaType} this is the name of the document
 * @param params {Record<string, unknown>} params is a sanity context object passed to every initial value function
 * @param maxDepth {Record<string, unknown>} maximum recursion depth (default 9)
 */
export function resolveInitialValueForType(
  type: SchemaType,
  params: InitialValueParams = {},
  maxDepth = MAX_RECURSION_DEPTH
) {
  if (maxDepth <= 0) {
    return undefined
  }
  if (isObjectSchemaType(type)) {
    return resolveInitialObjectValue(type, params, maxDepth)
  }
  if (isArraySchemaType(type)) {
    return resolveInitialArrayValue(type, params, maxDepth)
  }
  return resolveValue(type.initialValue, params)
}

async function resolveInitialArrayValue(type, params: InitialValueParams, maxDepth: number) {
  const initialArray = await resolveValue(type.initialValue)
  return Array.isArray(initialArray)
    ? Promise.all(
        initialArray.map(async (initialItem) => {
          const itemType = getItemType(type, initialItem)!
          return isObjectSchemaType(itemType)
            ? {
                ...initialItem,
                ...(await resolveInitialValueForType(itemType, params, maxDepth - 1)),
                _key: randomKey(),
              }
            : initialItem
        })
      )
    : undefined
}
export async function resolveInitialObjectValue(
  type: ObjectSchemaType,
  params: InitialValueParams,
  maxDepth: number
) {
  const initialObject: Record<string, unknown> = {
    ...((await resolveValue(type.initialValue, params)) || {}),
  }

  const fieldValues = {}
  await Promise.all(
    type.fields.map(async (field) => {
      const initialFieldValue = await resolveInitialValueForType(field.type, params, maxDepth - 1)
      if (initialFieldValue !== undefined && initialFieldValue !== null) {
        fieldValues[field.name] = initialFieldValue
      }
    })
  )

  const merged = deepAssign(fieldValues, initialObject)
  if (isEmpty(merged)) {
    return undefined
  }
  if (type.name !== 'object') {
    merged._type = type.name
  }
  return merged
}
