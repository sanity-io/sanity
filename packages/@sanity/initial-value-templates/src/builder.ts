import {Template, TemplateBuilder} from './Template'
import {Schema, SchemaType, getDefaultSchema} from './parts/Schema'

function defaultTemplateForType(
  schemaType: string | SchemaType,
  sanitySchema?: Schema
): TemplateBuilder {
  let type: SchemaType
  if (typeof schemaType === 'string') {
    const schema = sanitySchema || getDefaultSchema()
    type = schema.get(schemaType)
  } else {
    type = schemaType
  }

  return new TemplateBuilder({
    id: type.name,
    schemaType: type.name,
    title: type.title || type.name,
    icon: type.icon,
    value: type.initialValue || {_type: type.name},
  })
}

function defaults(sanitySchema?: Schema) {
  const schema = sanitySchema || getDefaultSchema()
  if (!schema) {
    throw new Error(
      'Unable to automatically resolve schema. Pass schema explicitly: `defaults(schema)`'
    )
  }

  return schema
    .getTypeNames()
    .filter((typeName) => !/^sanity\./.test(typeName))
    .filter((typeName) => isDocumentSchemaType(typeName, schema))
    .map((typeName) => defaultTemplateForType(schema.get(typeName), schema))
}

function isDocumentSchemaType(typeName: string, schema: Schema) {
  const schemaType = schema.get(typeName)
  return schemaType.type && schemaType.type.name === 'document'
}

export default {
  template: (spec?: Template) => new TemplateBuilder(spec),
  defaults,
  defaultTemplateForType,
}
