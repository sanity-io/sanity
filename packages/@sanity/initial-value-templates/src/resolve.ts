import {isPlainObject} from 'lodash'
import {Template, TemplateBuilder} from './Template'
import {validateInitialValue} from './validate'

export function isBuilder(template: Template | TemplateBuilder): template is TemplateBuilder {
  return typeof (template as TemplateBuilder).serialize === 'function'
}

async function resolveInitialValue(
  template: Template | TemplateBuilder,
  params: {[key: string]: any} = {}
): Promise<{[key: string]: any}> {
  // Template builder?
  if (isBuilder(template)) {
    return resolveInitialValue(template.serialize(), params)
  }

  const {id, value} = template
  if (!value) {
    throw new Error(`Template "${id}" has invalid "value" property`)
  }

  // Static value?
  if (isPlainObject(value)) {
    return validateInitialValue(value, template)
  }

  // Not an object, so should be a function
  if (typeof value !== 'function') {
    throw new Error(
      `Template "${id}" has invalid "value" property - must be a plain object or a resolver function`
    )
  }

  const resolved = await value(params)
  return validateInitialValue(resolved, template)
}

export {resolveInitialValue}
