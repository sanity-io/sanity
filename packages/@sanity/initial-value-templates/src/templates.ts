/* eslint-disable import/prefer-default-export */
import T from './builder'
import {validateTemplates} from './validate'
import {Template, TemplateBuilder} from './Template'
import {Schema} from './parts/Schema'
import {isBuilder} from './resolve'

interface CommonJsEsStub {
  __esModule: boolean
  default: (Template | TemplateBuilder)[]
}

function isCommonJsEsStub(mod: any): mod is CommonJsEsStub {
  const stub = mod as CommonJsEsStub
  return stub.__esModule && 'default' in stub
}

function loadDefinedTemplatesFromPart() {
  try {
    const templates:
      | undefined
      | (Template | TemplateBuilder)[]
      | CommonJsEsStub = require('part:@sanity/base/initial-value-templates?')

    if (typeof templates === 'undefined') {
      return undefined
    }

    if (!templates) {
      return null
    }

    if (isCommonJsEsStub(templates)) {
      return templates.default
    }

    return templates
  } catch (err) {
    return undefined
  }
}

function loadDefinedTemplates(schema?: Schema) {
  let templates = loadDefinedTemplatesFromPart()

  // Templates is `null` if the part was implemented but the export doesn't make sense,
  // if it is not found (require failed), the value will be `undefined`
  if (templates === null || (templates && !Array.isArray(templates))) {
    const type = templates === null ? `null` : typeof templates
    throw new Error(
      `'part:@sanity/base/initial-value-templates' should be an array of templates, got ${type}`
    )
  }

  if (!templates) {
    templates = T.defaults(schema)
  }

  return prepareTemplates(templates)
}

function prepareTemplates(templates: (Template | TemplateBuilder)[]) {
  const serialized = templates.map(maybeSerialize)
  return validateTemplates(serialized)
}

function getDefaultTemplates(schema?: Schema) {
  return prepareTemplates(T.defaults(schema))
}

function maybeSerialize(template: Template | TemplateBuilder) {
  return isBuilder(template) ? template.serialize() : template
}

export function getTemplateErrors(schema: Schema) {
  try {
    loadDefinedTemplates(schema)
    return []
  } catch (err) {
    return [err]
  }
}

export function getTemplates(schema?: Schema) {
  try {
    return loadDefinedTemplates(schema)
  } catch (err) {
    // eslint-disable-next-line no-console
    console.warn('Failed to load defined templates, falling back to defaults:\n%s', err.message)
    return getDefaultTemplates(schema)
  }
}

export function getParameterlessTemplatesBySchemaType(schemaType: string) {
  return getTemplatesBySchemaType(schemaType).filter(
    tpl => !tpl.parameters || !tpl.parameters.length
  )
}

export function getTemplatesBySchemaType(schemaType: string) {
  return getTemplates().filter(tpl => tpl.schemaType === schemaType)
}

export function getTemplateById(id: string) {
  return getTemplates().find(tpl => tpl.id === id)
}

export function templateExists(id: string) {
  return Boolean(getTemplateById(id))
}
