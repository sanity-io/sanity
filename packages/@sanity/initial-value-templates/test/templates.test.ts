import {format} from 'util'
import {getTemplates} from '../src'

beforeEach(() => {
  jest.resetModules()
})

function getConsoleSpyWithAssert() {
  const consoleSpy = jest.spyOn(global.console, 'warn')
  consoleSpy.mockImplementation((msg, ...args) =>
    expect(format(msg, ...args)).toMatchSnapshot('warning')
  )
  return {restore: () => consoleSpy.mockRestore()}
}

describe('getTemplates', () => {
  test('returns defaults if part is not implemented', () => {
    expect(getTemplates()).toMatchSnapshot()
  })

  test('returns defined templates if part implemented', () => {
    jest.mock('part:@sanity/base/initial-value-templates?', () => [
      {
        id: 'author',
        title: 'Author',
        schemaType: 'author',
        value: {title: 'here'}
      },
      {
        serialize: () => ({
          id: 'developer',
          title: 'Developer',
          schemaType: 'developer',
          value: {title: 'Foo'}
        })
      }
    ])
    expect(getTemplates()).toMatchSnapshot()
  })

  test('validates that templates has ID', () => {
    jest.mock('part:@sanity/base/initial-value-templates?', () => [
      {
        title: 'Author',
        schemaType: 'author',
        value: {title: 'here'}
      }
    ])

    const consoleSpy = getConsoleSpyWithAssert()
    expect(getTemplates()).toMatchSnapshot()
    consoleSpy.restore()
  })

  test('validates that templates has title', () => {
    jest.mock('part:@sanity/base/initial-value-templates?', () => [
      {
        id: 'author',
        schemaType: 'author',
        value: {title: 'here'}
      }
    ])

    const consoleSpy = getConsoleSpyWithAssert()
    expect(getTemplates()).toMatchSnapshot()
    consoleSpy.restore()
  })

  test('validates that templates has schema type', () => {
    jest.mock('part:@sanity/base/initial-value-templates?', () => [
      {
        id: 'author',
        title: 'Author',
        value: {title: 'here'}
      }
    ])

    const consoleSpy = getConsoleSpyWithAssert()
    expect(getTemplates()).toMatchSnapshot()
    consoleSpy.restore()
  })

  test('validates that templates has value', () => {
    jest.mock('part:@sanity/base/initial-value-templates?', () => [
      {
        id: 'author',
        title: 'Author',
        schemaType: 'author'
      }
    ])

    const consoleSpy = getConsoleSpyWithAssert()
    expect(getTemplates()).toMatchSnapshot()
    consoleSpy.restore()
  })

  test('validates that templates has id, title, schemaType, value', () => {
    jest.mock('part:@sanity/base/initial-value-templates?', () => [{}])

    const consoleSpy = getConsoleSpyWithAssert()
    expect(getTemplates()).toMatchSnapshot()
    consoleSpy.restore()
  })

  test('validates that templates has an object/function value', () => {
    jest.mock('part:@sanity/base/initial-value-templates?', () => [
      {
        id: 'author',
        title: 'Author',
        schemaType: 'author',
        value: []
      }
    ])

    const consoleSpy = getConsoleSpyWithAssert()
    expect(getTemplates()).toMatchSnapshot()
    consoleSpy.restore()
  })

  test('validates that templates has unique IDs', () => {
    jest.mock('part:@sanity/base/initial-value-templates?', () => [
      {
        id: 'author',
        title: 'Author',
        schemaType: 'author',
        value: {name: 'Gunnar'}
      },
      {
        id: 'author',
        title: 'Developer',
        schemaType: 'author',
        value: {role: 'developer'}
      }
    ])

    const consoleSpy = getConsoleSpyWithAssert()
    expect(getTemplates()).toMatchSnapshot()
    consoleSpy.restore()
  })
})
