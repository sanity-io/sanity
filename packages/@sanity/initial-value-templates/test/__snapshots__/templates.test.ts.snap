// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTemplates returns defaults if part is not implemented 1`] = `
Array [
  Object {
    "description": undefined,
    "icon": [Function],
    "id": "author",
    "parameters": undefined,
    "schemaType": "author",
    "title": "Author",
    "value": [Function],
  },
  Object {
    "description": undefined,
    "icon": [Function],
    "id": "post",
    "parameters": undefined,
    "schemaType": "post",
    "title": "Post",
    "value": Object {
      "_type": "post",
    },
  },
]
`;

exports[`getTemplates returns defined templates if part implemented 1`] = `
Array [
  Object {
    "id": "author",
    "schemaType": "author",
    "title": "Author",
    "value": Object {
      "title": "here",
    },
  },
  Object {
    "id": "developer",
    "schemaType": "developer",
    "title": "Developer",
    "value": Object {
      "title": "Foo",
    },
  },
]
`;

exports[`getTemplates validates that templates has ID 1`] = `
Array [
  Object {
    "description": undefined,
    "icon": [Function],
    "id": "author",
    "parameters": undefined,
    "schemaType": "author",
    "title": "Author",
    "value": [Function],
  },
  Object {
    "description": undefined,
    "icon": [Function],
    "id": "post",
    "parameters": undefined,
    "schemaType": "post",
    "title": "Post",
    "value": Object {
      "_type": "post",
    },
  },
]
`;

exports[`getTemplates validates that templates has ID: warning 1`] = `
"Failed to load defined templates, falling back to defaults:
Template \\"Author\\" is missing required properties: id"
`;

exports[`getTemplates validates that templates has an object/function value 1`] = `
Array [
  Object {
    "description": undefined,
    "icon": [Function],
    "id": "author",
    "parameters": undefined,
    "schemaType": "author",
    "title": "Author",
    "value": [Function],
  },
  Object {
    "description": undefined,
    "icon": [Function],
    "id": "post",
    "parameters": undefined,
    "schemaType": "post",
    "title": "Post",
    "value": Object {
      "_type": "post",
    },
  },
]
`;

exports[`getTemplates validates that templates has an object/function value: warning 1`] = `
"Failed to load defined templates, falling back to defaults:
Template \\"author\\" has an invalid \\"value\\" property; should be a function or an object"
`;

exports[`getTemplates validates that templates has id, title, schemaType, value 1`] = `
Array [
  Object {
    "description": undefined,
    "icon": [Function],
    "id": "author",
    "parameters": undefined,
    "schemaType": "author",
    "title": "Author",
    "value": [Function],
  },
  Object {
    "description": undefined,
    "icon": [Function],
    "id": "post",
    "parameters": undefined,
    "schemaType": "post",
    "title": "Post",
    "value": Object {
      "_type": "post",
    },
  },
]
`;

exports[`getTemplates validates that templates has id, title, schemaType, value: warning 1`] = `
"Failed to load defined templates, falling back to defaults:
Template at index 0 is missing required properties: id, title, schemaType, value"
`;

exports[`getTemplates validates that templates has schema type 1`] = `
Array [
  Object {
    "description": undefined,
    "icon": [Function],
    "id": "author",
    "parameters": undefined,
    "schemaType": "author",
    "title": "Author",
    "value": [Function],
  },
  Object {
    "description": undefined,
    "icon": [Function],
    "id": "post",
    "parameters": undefined,
    "schemaType": "post",
    "title": "Post",
    "value": Object {
      "_type": "post",
    },
  },
]
`;

exports[`getTemplates validates that templates has schema type: warning 1`] = `
"Failed to load defined templates, falling back to defaults:
Template \\"author\\" is missing required properties: schemaType"
`;

exports[`getTemplates validates that templates has title 1`] = `
Array [
  Object {
    "description": undefined,
    "icon": [Function],
    "id": "author",
    "parameters": undefined,
    "schemaType": "author",
    "title": "Author",
    "value": [Function],
  },
  Object {
    "description": undefined,
    "icon": [Function],
    "id": "post",
    "parameters": undefined,
    "schemaType": "post",
    "title": "Post",
    "value": Object {
      "_type": "post",
    },
  },
]
`;

exports[`getTemplates validates that templates has title: warning 1`] = `
"Failed to load defined templates, falling back to defaults:
Template \\"author\\" is missing required properties: title"
`;

exports[`getTemplates validates that templates has unique IDs 1`] = `
Array [
  Object {
    "description": undefined,
    "icon": [Function],
    "id": "author",
    "parameters": undefined,
    "schemaType": "author",
    "title": "Author",
    "value": [Function],
  },
  Object {
    "description": undefined,
    "icon": [Function],
    "id": "post",
    "parameters": undefined,
    "schemaType": "post",
    "title": "Post",
    "value": Object {
      "_type": "post",
    },
  },
]
`;

exports[`getTemplates validates that templates has unique IDs: warning 1`] = `
"Failed to load defined templates, falling back to defaults:
Template \\"Developer\\" at index 1 has the same ID (\\"author\\") as template \\"Author\\" at index 0"
`;

exports[`getTemplates validates that templates has value 1`] = `
Array [
  Object {
    "description": undefined,
    "icon": [Function],
    "id": "author",
    "parameters": undefined,
    "schemaType": "author",
    "title": "Author",
    "value": [Function],
  },
  Object {
    "description": undefined,
    "icon": [Function],
    "id": "post",
    "parameters": undefined,
    "schemaType": "post",
    "title": "Post",
    "value": Object {
      "_type": "post",
    },
  },
]
`;

exports[`getTemplates validates that templates has value: warning 1`] = `
"Failed to load defined templates, falling back to defaults:
Template \\"author\\" is missing required properties: value"
`;
